Sentralisert fil for import av funksjonar

haugedal opened the pull request.

haugedal committed:
Bruk sentralisert fil for import av funksjonar

haugedal committed:
Fjerna import av heile pakkar

Og lagt til manglande import av enkeltfunksjonar i den
sentraliserte fila. Me kan eventuelt vurdera om det er
greitt å importera heile dplyr-pakken.

haugedal commented:
Me får diskutera om enkelte pakkar (f.eks. `dplyr` og `ggplot2`?) skal importerast heilt, og om andre mindre brukte funksjonar heller skal bruka `::` enn `@importFrom`. Her er i alle fall ein start, og så er det enklare å få ei oversikt over importerte funksjonar slik dei ligg i greina her no.

[roxygen2](https://roxygen2.r-lib.org/articles/namespace.html#imports) skriv om bruk av `usethis::use_import_from()` til å automatisera import av funksjonar. Eg har brukt denne funksjonen no, men han puttar kvar funksjon på ei eiga `@importFrom`-linje, og samlar altså ikkje alle funksjonar frå éin pakke på éi linje. Dette kan me òg sjå nærmare på, og diskutera korleis me vil ha det. Ser det er gjort litt ulikt for dei ulike `tidyverse`-pakkane, men alle har i alle fall éi sentralisert fil for import. Ein del av dei importerer òg heile pakkar, blant anna er heile `rlang` importert i fleire andre pakkar.

statkarl commented:
Eg stemmer for å *ikkje* importera heile pakkar. Andre ting:

- Eg ser det er import av sjeldne brukte funksjonar, eks. `@importFrom styler tidyverse_math_token_spacing`. Det er vel ein uting. Her må det vera betre å bruka `styler::tidyverse_math_token_spacing` den *eine* plassen denne obskure funksjonen vert brukt. Me bør berre importera hyppig brukte og godt kjende funksjonar. Då vert òg importeringsfila meir oversiktleg.
- Korleis føregår vedlikehaldet av fila? Altså, korleis skal me hugsa å fjerna funksjonar av fila når me ikkje lenger brukar dei? (Eitt eksempel er ruler-pakken, som me kanskje vil kvitta oss med ein eller annan gong.) Er det noko som åtvarar oss (eller helst gjev feilmelding) dersom me importerer ubrukte funksjonar? (Det vert kanskje uansett eit mindre problem dersom me berre har hyppig brukte og godt kjende funksjonar i fila.)
- Me bør vel òg få noko inn i stilgaiden om dette? Dokumentera korleis me gjer ting.

statkarl commented on DESCRIPTION:
```diff
@@ -50,7 +50,8 @@ Imports:
     MASS,
-    boot
+    boot,
+    tidyselect
```
`tidyselect` mangla. Korfor har me ikkje fått feilmelding på dette tidlegare?

haugedal replied:
Dei to aktuelle `tidyselect`-funksjonane `all_of()` og `where()` er importert i `dplyr`, og me hadde tilgang til dei ved å importerta heile `dplyr`-pakken.

statkarl commented:
Elles so synest eg det er fint at kvar import står på éi linje (og alfabetisk, og gruppert og sortert alfabetisk etter pakke). Det gjev òg finare diffar.

statkarl commented on R/rapwhale-package.R:
```diff
@@ -0,0 +1,127 @@
+#' @importFrom dplyr group_vars
+#' @importFrom dplyr groups
+#' @importFrom dplyr if_else
+#' @importFrom dplyr intersect
```
Har me eksempel på at me brukar `intersect()` frå dplyr (i staden for frå base)?

haugedal replied:
Nei, det har me kanskje ikkje, så den kan kanskje fjernast. Eg tenkte berre at sidan me tidlegare hadde importert heile `dplyr`-pakken, så la eg til denne linja for at det skulle vera identisk som tidlegare.

statkarl replied:
Bør fjernast. Me brukar berre `intersect()` på vektorar, ikkje datarammer (som varianten i dplyr er laga for).

haugedal replied:
Fjerna.

statkarl commented on R/rapwhale-package.R:
```diff
@@ -0,0 +1,127 @@
+#' @importFrom dplyr select
+#' @importFrom dplyr slice
+#' @importFrom dplyr summarise
+#' @importFrom dplyr summarise_at
+#' @importFrom dplyr tally
+#' @importFrom dplyr transmute_at
```
Desse `*_at()`-funksjonane skal me jo kvitta oss med.

haugedal replied:
Ja, var her eg kom over at dei var i bruk. Kva som skjer fyrst av fjerning av desse funksjonane, eller bruk av sentralisert importfil er vel ikkje så viktig (berre me hugsar å fjerna dei frå denne fila når det vert gjort).

statkarl replied:
Eg har no lagt til eit sjekklistepunkt på tilhøyrande Trello-kort om fjerning av linjene (https://trello.com/c/CHlrNYm5/2106-oppdatere-scoped-verbs-i-rapwhale-til-%C3%A5-heller-bruke-across).

statkarl commented on R/rapwhale-package.R:
```diff
@@ -0,0 +1,127 @@
+#' @importFrom purrr possibly
+#' @importFrom purrr walk
+#' @importFrom qicharts2 qic
+#' @importFrom Rdpack reprompt
```
Denne vert aldri brukt til noko?

haugedal replied:
Viss eg forstod det rett så har denne noko med [import av `BibTeX`-referansar](https://geobosh.github.io/Rdpack/reference/Rdpack-package.html) å gjera, og skal stå slik.

statkarl replied:
Okei. Då gav det meir meining at han før stod i fila `R/regn-sumskaar-rand12.R` (sidan hjelpefila der har ein referanse til ein vitskapleg artikkel).

haugedal replied:
Sjå 6d8df76d6c651b3ce51fb6c3ac73785b53cfb7e2 og 67bee709ea33eeb282d51b2b599d640f90b9b423.

haugedal replied:
Eg kommenterte visst her før eg såg den nye kommentaren din.

haugedal commented:
Einig i dei overordna kommentarane dine.

Tenkjer at dette gav ei grei oversikt over importerte funksjonar, og så kan me som du føreslår gå over til bruk av `::` for sjeldan brukte funksjonar.

Når det gjeld vedlikehald av fila så syntest eg i alle fall at `usethis::use_import_from()` var fin for å leggja til nye funksjonar. Då kjem dei automatisk i alfabetisk rekkjefylgje, og det skjer heller ingenting om ein prøver å leggja til ein som allerede er importert. For sletting av ubrukte funksjonar kan me jo eventuelt sjå litt på kva alternativ som finst for automatisering. Elles må det vel berre gjerast manuelt, men eg trur dei automatiske sjekkane me har på GitHub vil klaga på dette, i alle fall med ei åtvaring. Eventuelt kan me leggja inn eit punkt på å sjekka dette i pakkeutleveringsrutinane?

Ja, me bør få inn noko i stilguiden om dette. Me har eit kort på det i Trello.

statkarl commented:
Kan kanskje laga eit kjapt skript som lagar statistikk på kor ofte (dvs. på kor mange linjer) kvar av dei importerte funksjonane vert brukte? Så kan me utelukka dei minst brukte.

haugedal commented:
Ja. Og same skriptet kan jo eventuelt òg nyttast til å sjekka at alle importerte funksjonar faktisk er i bruk.

haugedal committed:
Ikkje importer intersect() frå dplyr-pakken

Kall funksjonen eksplisitt med :: dersom
dplyr-versjonen av intersect() skal brukast.

statkarl reviewed (approved):
LGTM
